onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,721 @@
+    // NodeArg is zero padded to the NCHWc block size. If the output needs to
+    // be reordered back to a standard tensor format, this channel count is
+    // used to generate the expected number of channels.
+    const int64_t channels_;
+
+    // Stores the proto shape for the NCHWc output.
+    NchwcArgument::Shape shape_;
+
+    NchwcArgument(Node& output_node, NodeArg* output_nchwc_arg, size_t original_uses, size_t channels, const NchwcArgument::Shape& shape)
+        : output_node_(output_node),
+          nchwc_arg_(output_nchwc_arg),
+          starting_original_uses_(original_uses),
+          remaining_original_uses_(original_uses),
+          channels_(channels),
+          shape_(shape) {
+    }
+  };
+
+  const ONNX_NAMESPACE::AttributeProto* GetAttribute(const Node& node, const char* attribute_name);
+  size_t RemoveOutputEdges(Node& node);
+  void CreateNchwcArgument(Node& node, Node& nchwc_node, int64_t channels, const NchwcArgument::Shape& shape);
+  void FuseNchwcArgument(Node& node, const NchwcArgument& nchwc_arg);
+  void InsertReorderInput(Node& node);
+
+  void ConvPoolShapeInference(const Node& node,
+                              const NchwcArgument::Shape& input_shape,
+                              NchwcArgument::Shape& output_shape,
+                              const ONNX_NAMESPACE::TensorProto* filter_shape);
+
+  void TransformConv(Node& node);
+  void TransformPool(Node& node);
+  void TransformAdd(Node& node);
+  void TransformConcat(Node& node);
+  void TransformActivation(Node& node);
+  void TransformElementwise(Node& node);
+
+  Graph& graph_;
+
+  // Stores a queue of nodes to be removed after walking through the graph.
+  std::deque<NodeIndex> removed_nodes_;
+
+  // Stores a mapping from the original NodeArg outputs to the NCHWc variants
+  // created inside this graph transform.
+  std::unordered_map<NodeArg*, std::unique_ptr<NchwcArgument>> nchwc_args_;
+
+  // Stores a mapping of NodeArg inputs that have already been reordered, so
+  // multiple nodes can share the NCHWc input.
+  std::unordered_map<NodeArg*, NodeArg*> reorder_inputs_;
+
+  // Stores a mapping of NodeArg filters that have already been reordered, so
+  // multiple nodes can share the NCHWc filter.
+  std::unordered_map<NodeArg*, NodeArg*> filters_OIHWBo_;
+  std::unordered_map<NodeArg*, NodeArg*> filters_OIHWBiBo_;
+
+  // Stores a mapping of NodeArg biases that have already been aligned to the
+  // NCHWc block size, so multiple nodes can share the NCHWc biases.
+  std::unordered_map<NodeArg*, NodeArg*> aligned_biases_;
+};
+
+const ONNX_NAMESPACE::AttributeProto* NchwcTransformerImpl::GetAttribute(const Node& node, const char* attribute_name) {
+  auto& node_attributes = node.GetAttributes();
+  auto it = node_attributes.find(attribute_name);
+  if (it != node_attributes.end()) {
+    return &(it->second);
+  } else {
+    return nullptr;
+  }
+}
+
+size_t NchwcTransformerImpl::RemoveOutputEdges(Node& node) {
```

Comment:
Is this generic enough to be in graph_utils?

Reply:
Maybe something for the optimizer team to do if they like, but I would prefer not to deal with this in this PR.

Reply:
Shall we use the graph_utils::RemoveNodeOutputEdges() here? At least for biggest part of the method (I see there is the final else that might affect the returning value here).

Reply:
Fixed. I didn't like how RemoveNodeOutputEdges has grown to capture the nodearg name in a GraphEdge, but not worth fighting about now. :)